File: build-tools/android-builder/src/main/java/com/tyron/builder/android/aapt2/Aapt2Exception.java
package com.tyron.builder.android.aapt2;

/** Exception thrown when an error occurs during `aapt2` processing.  */
public class Aapt2Exception extends RuntimeException {

    private static final long serialVersionUID = 7034893190645766936L;

    public static Aapt2Exception create(
            String description
    ) {
        return new Aapt2Exception(description, null, null, null, null);
    }

    public static Aapt2Exception create(
            String description,
            Throwable cause,
            String output,
            String processName,
            String command
    ) {
        return new Aapt2Exception(description, cause, output, processName, command);
    }

    public Aapt2Exception(
            String description,
            Throwable cause,
            String output,
            String processName,
            String command
    ) {
        super(makeMessage(description, output), cause);
    }


    private static String makeMessage(String description, String output) {
        if (output == null) {
            return description;
        }
        return description + "\n" + output;
    }
}

File: build-tools/codeassist-builder-plugin/src/main/java/com/tyron/builder/internal/aapt/v2/Aapt2Exception.kt
package com.tyron.builder.internal.aapt.v2

import com.android.utils.ILogger

/** Exception thrown when an error occurs during `aapt2` processing.  */
class Aapt2Exception private constructor(
    val description: String,
    cause: Throwable? = null,
    val output: String? = null,
    val processName: String? = null,
    val command: String? = null
 ) : RuntimeException(makeMessage(description, output), cause) {
    companion object {
        private const val serialVersionUID = 7034893190645766936L
        @JvmStatic
        @JvmOverloads
        fun create(
            logger: ILogger? = null,
            description: String,
            cause: Throwable? = null,
            output: String? = null,
            processName: String? = null,
            command: String? = null

        ) : Aapt2Exception {
            logger?.info("$description:\n" +
                    "process = $processName:\n" +
                    "command = $command")
            return Aapt2Exception(
                description = description,
                cause = cause,
                output = output,
                processName = processName,
                command = command)
        }
    }
}
private fun makeMessage(description: String, output: String?):String {
    if (output == null) {
        return description
    }
    return description + "\n" + output
}
